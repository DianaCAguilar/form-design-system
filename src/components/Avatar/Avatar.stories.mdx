import { Meta, Story, Preview, Props } from "@storybook/addon-docs/blocks";
import { withKnobs, text, boolean, optionsKnob as options } from '@storybook/addon-knobs';
import StarFilledIcon from 'lib/icons/react/StarFilledIcon';
import ImportPath from 'util/storybook-docs/ImportPath';
import { arrayToOptions } from 'util/storybook-docs/knobHelpers';
import { StoryWrapper, StoryItem } from 'util/storybook-docs/StoryLayout';
import { action } from '@storybook/addon-actions';

import Avatar, { BG_COLORS, SIZES, RADII, INITIALS_LENGTH } from '.';

<Meta title="Components/Avatar" decorators={[withKnobs]} component={Avatar} />

# Avatar

<ImportPath component={Avatar} />

<Story name="Knobs" parameters={{ docs: { disable: true }}}>
  <StoryWrapper>
    <StoryItem>
      <Avatar
        bgColor={options('bgColor', arrayToOptions(BG_COLORS), undefined, {
          display: 'inline-radio',
        })}
        radius={options('radius', arrayToOptions(RADII), undefined, {
          display: 'inline-radio',
        })}
        size={options('size', arrayToOptions(SIZES), undefined, {
          display: 'inline-radio',
        })}
        name={text('name', 'Samuel Jackson')}
        imgUrl={text('imgUrl', undefined)}
        initialsLength={options('initials', arrayToOptions(INITIALS_LENGTH), undefined, {
          display: 'inline-radio',
        })}
        PlaceholderIcon={boolean("PlaceholderIcon", false) ? StarFilledIcon : undefined}
        href={text('href', '')}
      />
    </StoryItem>
    <StoryItem>
      <Avatar
        bgColor={options('bgColor', arrayToOptions(BG_COLORS), undefined, {
          display: 'inline-radio',
        })}
        radius={options('radius', arrayToOptions(RADII), undefined, {
          display: 'inline-radio',
        })}
        size={options('size', arrayToOptions(SIZES), undefined, {
          display: 'inline-radio',
        })}
        name={text('name', 'Samuel Jackson')}
        imgUrl="http://placeimg.com/80/81/animals"
        initialsLength={options('initials', arrayToOptions(INITIALS_LENGTH), undefined, {
          display: 'inline-radio',
        })}
        PlaceholderIcon={boolean("PlaceholderIcon", false) ? StarFilledIcon : undefined}
        href={text('href', '')}
      />
    </StoryItem>
  </StoryWrapper>
</Story>

<Preview isExpanded>
  <Avatar name='Samuel Jackson' />
</Preview>


_Displaying an Avatar with an image, and without an image, for visualization purposes_

<Props of={Avatar} />

### Product Cases

_Pay note to `initialsLength` in these cases._

<Preview>
  <Story name="Product Cases">
    <StoryWrapper>
      <StoryItem>
        <p>Entity Avatar</p>
      </StoryItem>
      <StoryItem>
        <Avatar name="Hello Impecable" bgColor="navy" initialsLength={1} radius="square" />
      </StoryItem>
      <StoryItem>
        <p>User Avatar</p>
      </StoryItem>
      <StoryItem>
        <Avatar name="Hello Impecable" />
      </StoryItem>
    </StoryWrapper>
  </Story>
</Preview>

### Colors
Accepted colors are based off FDS `bgColor--{color}` classes: https://cbinsights.github.io/form-design-system/fds-styles/#background-colors

<Preview>
  <Story name="Colors">
    <StoryWrapper>
      <StoryItem>
        <Avatar name="default" />
      </StoryItem>
      {BG_COLORS.map(bgColor => (
        <StoryItem>
          <Avatar name={`${bgColor.slice(0,1)} ${bgColor.slice(1,2)}`} bgColor={bgColor} />
        </StoryItem>
      ))}
      <br />
    </StoryWrapper>
  </Story>
</Preview>

### Images
Initials will show either if you don't pass an image, the image doesn't load, or it hasn't loaded yet.

<Preview>
  <Story name="Variations">
    <StoryWrapper>
      <StoryItem>
        <Avatar name="Hello Impecable" imgUrl="http://placeimg.com/80/81/animals" />
      </StoryItem>
      <StoryItem>
        <Avatar name="Hello Impecable" imgUrl="invalid url" />
      </StoryItem>
      <br />
    </StoryWrapper>
  </Story>
</Preview>

### Sizes

<Preview>
  <Story name="Sizes">
    <StoryWrapper>
      {SIZES.map(size => (
        <StoryItem>
          <Avatar name="Hello Impecable " size={size} />
        </StoryItem>
      ))}
      <br />
    </StoryWrapper>
  </Story>
</Preview>

### Radii

<Preview>
  <Story name="Radii">
    <StoryWrapper>
      {RADII.map(radius => (
        <StoryItem>
          <Avatar name="Hello Impecable " radius={radius} />
        </StoryItem>
      ))}
    </StoryWrapper>
  </Story>
</Preview>

### Button

Avatar will render as a button if you pass it a href, onClick, or Link to it.

<Preview>
  <Story name="Button">
    <StoryWrapper>
      <StoryItem>
        <Avatar name="Hello Impecable" href="#" />
      </StoryItem>
      <StoryItem>
        <Avatar name="Hello Impecable" onClick={action('avatar-click')} />
      </StoryItem>
      <StoryItem>
        <Avatar name="Hello Impecable" imgUrl="http://placeimg.com/80/81/animals" href="#" />
      </StoryItem>
    </StoryWrapper>
  </Story>
</Preview>

### Placeholder

<Preview>
  <Story name="Placeholder">
    <StoryWrapper>
      <StoryItem>
        <Avatar PlaceholderIcon={StarFilledIcon} />
      </StoryItem>
    </StoryWrapper>
  </Story>
</Preview>



