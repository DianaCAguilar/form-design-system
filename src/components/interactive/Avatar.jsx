import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import baseElement from '../../util/baseElement';

export const VALID_BG_COLORS = [
  'white',
  'haze',
  'lightGray',
  'orange',
  'charcoal',
  'navy',
  'aqua',
  'gray',
  'red',
];

const INVERTED_BG_COLORS = ['gray', 'charcoal', 'navy', 'orange'];

const grabInitials = (str) =>
  str
    .split(' ')
    .splice(0, 2)
    .reduce((prev, curr) => prev + curr.charAt(0), '');

const Avatar = ({ bgColor, imgUrl, size, name, Link, ...rest }) => {
  const Element = baseElement({ href: rest.href, as: Link });
  return (
    <Element
      {...rest}
      title={name}
      className={cx('fdsAvatar', `bgColor--${bgColor}`, `fdsAvatar--${size}`, {
        'color--white': INVERTED_BG_COLORS.includes(bgColor),
      })}
    >
      {imgUrl && (
        <span className="fdsAvatar-img" style={{ backgroundImage: `url(${imgUrl})` }} />
      )}
      {name && grabInitials(name)}
    </Element>
  );
};

Avatar.defaultProps = {
  bgColor: 'gray',
  size: 'm',
};

Avatar.propTypes = {
  /**
   * Takes in a react-router `Link` reference and sets it
   * as the base element. You may ONLY use it like the
   * following:
   * ```
   * import { Link } from 'react-router'
   * Link={Link}
   * ```
   */
  Link: PropTypes.func,
  /** Controls color of button */
  bgColor: PropTypes.oneOf(VALID_BG_COLORS),
  /** Controls the size of the button */
  size: PropTypes.oneOf(['s', 'm']),
  /** Control initials displayed, and also used for title accessibility attribute */
  name: PropTypes.string.isRequired,
  /** Sets background image over initials */
  imgUrl: PropTypes.string,
};

export default Avatar;
