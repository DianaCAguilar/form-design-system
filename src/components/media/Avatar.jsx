import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import baseElement from '../../util/baseElement';

const LIGHT_COLORS = ['white', 'haze', 'lightGray'];
const DARK_COLORS = ['orange', 'charcoal', 'navy', 'aqua', 'gray', 'red', 'purple'];
export const RADII = ['square', 'circle'];
export const BG_COLORS = [...LIGHT_COLORS, ...DARK_COLORS];
export const SIZES = ['s', 'm'];

export const grabInitials = (str) =>
  str
    .split(' ')
    .slice(0, 2)
    .reduce((prev, curr) => prev + curr.charAt(0), '')
    .toUpperCase();

const Avatar = ({ bgColor, imgUrl, size, name, radius, Link, ...rest }) => {
  const Element = baseElement({ href: rest.href, onClick: rest.onClick, as: Link });
  return (
    <Element
      {...rest}
      role="img"
      title={name}
      className={cx(
        'fdsAvatar',
        `bgColor--${bgColor}`,
        `fdsAvatar--${size}`,
        `fdsAvatar--${radius}`,
        {
          'border--focus': Element !== 'div',
          'color--white': DARK_COLORS.includes(bgColor),
        }
      )}
    >
      {imgUrl && (
        <span className="fdsAvatar-img" style={{ backgroundImage: `url(${imgUrl})` }} />
      )}
      {name && grabInitials(name)}
    </Element>
  );
};

Avatar.defaultProps = {
  bgColor: 'purple',
  size: 'm',
  radius: 'circle',
};

Avatar.propTypes = {
  /** Controls color of button */
  bgColor: PropTypes.oneOf(BG_COLORS),
  /** Controls the size of the button */
  size: PropTypes.oneOf(SIZES),
  /** Controls radius of button (slightly rounded square, or circle) */
  radius: PropTypes.oneOf(RADII),
  /** Control initials displayed, and also used for title accessibility attribute (uses 2 words max for initials) */
  name: PropTypes.string.isRequired,
  /** Sets background image over initials */
  imgUrl: PropTypes.string,
  /**
   * Takes in a react-router `Link` reference and sets it
   * as the base element. You may ONLY use it like the
   * following:
   *
   * - `import { Link } from 'react-router'`
   * - `Link={Link}`
   */
  Link: PropTypes.func,
};

export default Avatar;
