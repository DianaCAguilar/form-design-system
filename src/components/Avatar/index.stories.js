import React from 'react';

import StarFilledIcon from 'lib/icons/react/StarFilledIcon';
import { StoryWrapper, StoryItem } from 'util/storybook-docs/StoryLayout';

import Avatar, { BG_COLORS, SIZES, RADII } from '.';

const Template = (args) => (
  <StoryWrapper>
    <StoryItem>
      <Avatar {...args} />
    </StoryItem>
    <StoryItem>
      <Avatar {...args} imgUrl="http://placeimg.com/80/81/animals" />
    </StoryItem>
  </StoryWrapper>
);

export const Primary = Template.bind({});
Primary.args = {
  name: 'Samuel Jackson',
};

export const PlaceholderIcon = Template.bind({});
PlaceholderIcon.args = {
  PlaceholderIcon: StarFilledIcon,
};

PlaceholderIcon.parameters = {
  docs: {
    description: {
      story: 'Renders if `name` and `imgUrl` are not present.',
    },
  },
};

export const Colors = () => (
  <StoryWrapper>
    <StoryItem>
      <Avatar name="default" />
    </StoryItem>
    {BG_COLORS.map((bgColor, idx) => (
      <StoryItem key={idx}>
        <Avatar
          name={`${bgColor.slice(0, 1)} ${bgColor.slice(1, 2)}`}
          bgColor={bgColor}
        />
      </StoryItem>
    ))}
  </StoryWrapper>
);
Colors.parameters = {
  docs: {
    description: {
      story:
        'Accepted colors are based off [background color classes](?path=/docs/style-utilities-background--background-colors).',
    },
  },
};

export const Images = () => (
  <StoryWrapper>
    <StoryItem>
      <Avatar name="Hello Impecable" imgUrl="http://placeimg.com/80/81/animals" />
    </StoryItem>
    <StoryItem>
      <Avatar name="Hello Impecable" imgUrl="invalid url" />
    </StoryItem>
  </StoryWrapper>
);
Images.parameters = {
  docs: {
    description: {
      story:
        "Initials will show either if you don't pass an image, the image doesn't load, or it hasn't loaded yet.",
    },
  },
};

export const Sizes = () => (
  <StoryWrapper>
    {SIZES.map((size, idx) => (
      <StoryItem key={idx}>
        <Avatar name="Hello Impecable " size={size} />
      </StoryItem>
    ))}
  </StoryWrapper>
);

export const Radii = () => (
  <StoryWrapper>
    {RADII.map((radius, idx) => (
      <StoryItem key={idx}>
        <Avatar name="Hello Impecable " radius={radius} />
      </StoryItem>
    ))}
  </StoryWrapper>
);

export const Button = () => (
  <StoryWrapper>
    <StoryItem>
      <Avatar name="Hello Impecable" href="#" />
    </StoryItem>
    <StoryItem>
      <Avatar name="Hello Impecable" />
    </StoryItem>
    <StoryItem>
      <Avatar
        name="Hello Impecable"
        imgUrl="http://placeimg.com/80/81/animals"
        href="#"
      />
    </StoryItem>
  </StoryWrapper>
);

Button.parameters = {
  docs: {
    description: {
      story:
        'Avatar will render as a button if you pass it a href, onClick, or Link to it.',
    },
  },
};

export default {
  component: Avatar,
  title: 'components/Avatar',
  parameters: {
    docs: {
      attributes: ['rest'],
      description: {
        component: '',
      },
    },
  },
};
