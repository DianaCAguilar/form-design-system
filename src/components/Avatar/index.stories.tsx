import React from 'react';
import { Story } from '@storybook/react';
import StarFilledIcon from 'icons/react/StarFilledIcon';
import { StoryWrapper, StoryItem } from 'util/storybook-docs/StoryLayout';

import Avatar, { AvatarProps, BG_COLORS, SIZES, RADII } from '.';

const Template = (args: AvatarProps) => (
  <StoryWrapper>
    <StoryItem>
      <Avatar {...args} />
    </StoryItem>
    <StoryItem>
      <Avatar {...args} imgUrl="http://placeimg.com/80/81/animals" />
    </StoryItem>
  </StoryWrapper>
);

export const Primary: Story<AvatarProps> = Template.bind({});
Primary.args = {
  name: 'Samuel Jackson',
};

export const PlaceholderIcon: Story<AvatarProps> = Template.bind({});
PlaceholderIcon.args = {
  PlaceholderIcon: StarFilledIcon,
};

PlaceholderIcon.parameters = {
  docs: {
    description: {
      story: 'Renders if `name` and `imgUrl` are not present.',
    },
  },
};

export const Colors = (): JSX.Element => (
  <StoryWrapper>
    <StoryItem>
      <Avatar name="default" />
    </StoryItem>
    {BG_COLORS.map((bgColor, idx) => (
      <StoryItem key={idx}>
        <Avatar
          name={`${bgColor.slice(0, 1)} ${bgColor.slice(1, 2)}`}
          bgColor={bgColor}
        />
      </StoryItem>
    ))}
  </StoryWrapper>
);

Colors.parameters = {
  docs: {
    description: {
      story:
        'Accepted colors are based off [background color classes](?path=/docs/style-utilities-background--background-colors).',
    },
  },
};

export const Images = (): JSX.Element => (
  <StoryWrapper>
    <StoryItem>
      <Avatar name="Hello Impecable" imgUrl="http://placeimg.com/80/81/animals" />
    </StoryItem>
    <StoryItem>
      <Avatar name="Hello Impecable" imgUrl="invalid url" />
    </StoryItem>
  </StoryWrapper>
);
Images.parameters = {
  docs: {
    description: {
      story:
        "Initials will show either if you don't pass an image, the image doesn't load, or it hasn't loaded yet.",
    },
  },
};

export const Sizes = (): JSX.Element => (
  <StoryWrapper>
    {SIZES.map((size, idx) => (
      <StoryItem key={idx}>
        <Avatar name="Hello Impecable " size={size} />
      </StoryItem>
    ))}
  </StoryWrapper>
);

export const Radii = (): JSX.Element => (
  <StoryWrapper>
    {RADII.map((radius, idx) => (
      <StoryItem key={idx}>
        <Avatar name="Hello Impecable " radius={radius} />
      </StoryItem>
    ))}
  </StoryWrapper>
);

export const Button = (): JSX.Element => (
  <StoryWrapper>
    <StoryItem>
      <Avatar name="Hello Impecable" href="#" />
    </StoryItem>
    <StoryItem>
      <Avatar name="Hello Impecable" />
    </StoryItem>
    <StoryItem>
      <Avatar
        name="Hello Impecable"
        imgUrl="http://placeimg.com/80/81/animals"
        href="#"
      />
    </StoryItem>
  </StoryWrapper>
);

Button.parameters = {
  docs: {
    description: {
      story:
        'Avatar will render as a button if you pass it a href, onClick, or Link to it.',
    },
  },
};

export default {
  component: Avatar,
  title: 'components/Avatar',
  parameters: {
    docs: {
      fds: {
        attributes: ['rest'],
        related: ['IconButton', 'StackedButton'],
      },
    },
  },
};
