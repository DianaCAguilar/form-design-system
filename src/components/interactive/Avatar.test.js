import React from 'react';
import { shallow } from 'enzyme';

import Avatar from './Avatar';

const renderComponent = (props) => shallow(<Avatar {...props} />);

const props = {
  name: 'Ross Gellar', bgColor: 'aqua', size: 's', imgUrl: 'lol.png', 
};

describe('Avatar component', () => {

  it('matches snapshot (default props)', () => {
    const component = renderComponent({ name: 'Joey Tribbiani' });
    expect(component).toMatchSnapshot();
  });

  it('matches snapshot (set all props)', () => {
    const component = renderComponent(props);
    expect(component).toMatchSnapshot();
  });
});
