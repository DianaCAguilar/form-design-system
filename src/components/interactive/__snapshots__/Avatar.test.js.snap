// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar component matches snapshot (default props) 1`] = `
<button
  className="fdsAvatar bgColor--gray fdsAvatar--m color--white"
  title="Joey Tribbiani"
>
  JT
</button>
`;

exports[`Avatar component matches snapshot (set all props) 1`] = `
<MockLink
  className="fdsAvatar bgColor--aqua fdsAvatar--s"
  title="Ross Gellar"
>
  <span
    className="fdsAvatar-img"
    style={
      Object {
        "backgroundImage": "url(lol.png)",
      }
    }
  />
  RG
</MockLink>
`;
